File: .claude/hooks/analyze-observations.py
#!/usr/bin/env python3
"""
Autobot Observation Analyzer - Mine patterns from observations.jsonl.

Utility script for querying and analyzing observations logged during
autonomous development sessions.

Usage:
  python analyze-observations.py --type test_failure
  python analyze-observations.py --type pattern
  python analyze-observations.py --recent 10
  python analyze-observations.py --summary
"""
import json
import sys
import argparse
from pathlib import Path
from collections import Counter
from datetime import datetime


def load_observations(path: Path) -> list[dict]:
    """Load all observations from JSONL file."""
    observations = []
    if path.exists():
        with open(path, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if line:
                    try:
                        observations.append(json.loads(line))
                    except json.JSONDecodeError:
                        pass
    return observations


def filter_by_type(observations: list[dict], obs_type: str) -> list[dict]:
    """Filter observations by type."""
    return [o for o in observations if o.get('type') == obs_type]


def get_recent(observations: list[dict], count: int) -> list[dict]:
    """Get most recent observations."""
    return observations[-count:] if len(observations) > count else observations


def print_summary(observations: list[dict]) -> None:
    """Print summary statistics."""
    if not observations:
        print("No observations found.")
        return

    type_counts = Counter(o.get('type', 'unknown') for o in observations)
    file_counts = Counter(o.get('file', 'unknown') for o in observations if o.get('file'))

    print("=== OBSERVATION SUMMARY ===\n")
    print(f"Total observations: {len(observations)}")

    print("\nBy Type:")
    for obs_type, count in type_counts.most_common():
        print(f"  {obs_type}: {count}")

    print("\nMost Affected Files:")
    for file_path, count in file_counts.most_common(5):
        print(f"  {file_path}: {count}")

    # Time range
    timestamps = [o.get('timestamp') for o in observations if o.get('timestamp')]
    if timestamps:
        print(f"\nTime Range:")
        print(f"  First: {min(timestamps)}")
        print(f"  Last: {max(timestamps)}")


def print_observations(observations: list[dict]) -> None:
    """Print observations in readable format."""
    for i, obs in enumerate(observations, 1):
        print(f"\n--- Observation {i} ---")
        print(f"Type: {obs.get('type', 'unknown')}")
        print(f"Time: {obs.get('timestamp', 'unknown')}")
        if obs.get('file'):
            print(f"File: {obs.get('file')}")
        if obs.get('message'):
            print(f"Message: {obs.get('message')}")
        if obs.get('output_snippet'):
            print(f"Output: {obs.get('output_snippet')[:200]}...")


def main():
    parser = argparse.ArgumentParser(description='Analyze Autobot observations')
    parser.add_argument('--type', help='Filter by observation type')
    parser.add_argument('--recent', type=int, help='Show N most recent observations')
    parser.add_argument('--summary', action='store_true', help='Show summary statistics')
    parser.add_argument('--path', default='.autobot/observations.jsonl',
                        help='Path to observations file')

    args = parser.parse_args()

    obs_path = Path(args.path)
    observations = load_observations(obs_path)

    if args.type:
        observations = filter_by_type(observations, args.type)
        print(f"Filtered to {len(observations)} observations of type '{args.type}'")

    if args.recent:
        observations = get_recent(observations, args.recent)

    if args.summary:
        print_summary(observations)
    else:
        print_observations(observations)


if __name__ == "__main__":
    main()

File: .claude/hooks/analyze-observations.js
#!/usr/bin/env node
/**
 * Autobot Observation Analyzer - Mine patterns from observations.jsonl.
 *
 * Utility script for querying and analyzing observations logged during
 * autonomous development sessions.
 *
 * Usage:
 *   node analyze-observations.js --type test_failure
 *   node analyze-observations.js --recent 10
 *   node analyze-observations.js --summary
 */
import fs from 'fs';

function loadObservations(filePath) {
  const observations = [];
  if (!fs.existsSync(filePath)) return observations;

  const content = fs.readFileSync(filePath, 'utf-8');
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try {
      observations.push(JSON.parse(trimmed));
    } catch {
      // Skip invalid lines
    }
  }
  return observations;
}

function filterByType(observations, type) {
  return observations.filter(o => o.type === type);
}

function getRecent(observations, count) {
  return observations.slice(-count);
}

function printSummary(observations) {
  if (observations.length === 0) {
    console.log('No observations found.');
    return;
  }

  const typeCounts = {};
  const fileCounts = {};

  for (const obs of observations) {
    const type = obs.type || 'unknown';
    typeCounts[type] = (typeCounts[type] || 0) + 1;
    if (obs.file) {
      fileCounts[obs.file] = (fileCounts[obs.file] || 0) + 1;
    }
  }

  console.log('=== OBSERVATION SUMMARY ===\n');
  console.log(`Total observations: ${observations.length}`);

  console.log('\nBy Type:');
  const sortedTypes = Object.entries(typeCounts).sort((a, b) => b[1] - a[1]);
  for (const [type, count] of sortedTypes) {
    console.log(`  ${type}: ${count}`);
  }

  console.log('\nMost Affected Files:');
  const sortedFiles = Object.entries(fileCounts).sort((a, b) => b[1] - a[1]).slice(0, 5);
  for (const [file, count] of sortedFiles) {
    console.log(`  ${file}: ${count}`);
  }

  const timestamps = observations.map(o => o.timestamp).filter(Boolean);
  if (timestamps.length > 0) {
    timestamps.sort();
    console.log('\nTime Range:');
    console.log(`  First: ${timestamps[0]}`);
    console.log(`  Last: ${timestamps[timestamps.length - 1]}`);
  }
}

function printObservations(observations) {
  for (let i = 0; i < observations.length; i++) {
    const obs = observations[i];
    console.log(`\n--- Observation ${i + 1} ---`);
    console.log(`Type: ${obs.type || 'unknown'}`);
    console.log(`Time: ${obs.timestamp || 'unknown'}`);
    if (obs.file) console.log(`File: ${obs.file}`);
    if (obs.message) console.log(`Message: ${obs.message}`);
    if (obs.output_snippet) console.log(`Output: ${obs.output_snippet.slice(0, 200)}...`);
  }
}

function main() {
  const args = process.argv.slice(2);
  let type = null;
  let recent = null;
  let summary = false;
  let obsPath = '.autobot/observations.jsonl';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--type' && args[i + 1]) {
      type = args[++i];
    } else if (args[i] === '--recent' && args[i + 1]) {
      recent = parseInt(args[++i], 10);
    } else if (args[i] === '--summary') {
      summary = true;
    } else if (args[i] === '--path' && args[i + 1]) {
      obsPath = args[++i];
    }
  }

  let observations = loadObservations(obsPath);

  if (type) {
    observations = filterByType(observations, type);
    console.log(`Filtered to ${observations.length} observations of type '${type}'`);
  }

  if (recent) {
    observations = getRecent(observations, recent);
  }

  if (summary) {
    printSummary(observations);
  } else {
    printObservations(observations);
  }
}

main();
